lightwood/encoder/categorical/binary.py

```diff
@@ -1,76 +1,162 @@
     """ # noqa
-    def __init__(self, is_target=False, target_weights=None):
+    def __init__(
+        self,
+        is_target: bool = False,
+        target_weights: Dict[str, float] = None,
+    ):
         super().__init__(is_target)
-        self.map = {}
-        self.rev_map = {}
+        """
+        :param is_target: Whether encoder featurizes target column
+        :param target_weights: Percentage of total population represented by each category (from [0, 1]).
+        """  # noqa
+
+        self.map = {}  # category name -> index
+        self.rev_map = {}  # index -> category name
         self.output_size = 2
+        self.encoder_class_type = str
 
+        # Weight-balance info if encoder represents target
         self.target_weights = None
-        self.index_weights = None
+        self.inv_target_weights = None
         if self.is_target:
             self.target_weights = target_weights
 
-    def prepare(self, priming_data):
+    def prepare(self, priming_data: Iterable[str]):
+        """
+        Given priming data, create a map/inverse-map corresponding category name to index (and vice versa).
+
+        :param priming_data: Binary data to encode
+        """  # noqa
         if self.is_prepared:
             raise Exception('You can only call "prepare" once for a given encoder.')
 
-        for x in priming_data:
-            x = str(x)
-            if x not in self.map:
-                self.map[x] = len(self.map)
-                self.rev_map[len(self.rev_map)] = x
+        unq_cats = np.unique([i for i in priming_data if i is not None]).tolist()
+        self.map = {cat: indx for indx, cat in enumerate(unq_cats)}
+        self.rev_map = {indx: cat for cat, indx in self.map.items()}
 
-            if len(self.map) == 2:
-                break
+        # Enforce only binary; map must have exactly 2 classes.
+        if len(self.map) != 2:
+            raise ValueError('Issue with dtype; data has > 2 classes.')
 
+        # For target-only, report on relative weights of classes
         if self.is_target:
-            self.index_weights = [None, None]
-            for word in self.map:
-                if self.target_weights is not None:
-                    self.index_weights[self.map[word]] = 1 / self.target_weights[word]
-                else:
-                    self.index_weights[self.map[word]] = 1
 
-            self.index_weights = torch.Tensor(self.index_weights)
+            self.inv_target_weights = torch.Tensor([1, 1])  # Equally wt. both classes
+
+            # If target weights provided, weight by inverse
+            if self.target_weights is not None:
+
+                # Check target weights properly specified
+                if sum([np.isclose(i, 0) for i in self.target_weights.values()]) > 0:
+                    raise ValueError('Target weights cannot be 0')
```

Comment:
If I recall correctly, `target_weights` will multiply each element by the respective component in the loss function; if we have a 0, it's akin to entirely ignoring a loss term. Pytorch already has an 'ignore_index' arg for this in most loss functions, moreover, I'm not 100% sure what this would do to numerical stability while training.